embassy-nrf/src/twis.rs

```diff
@@ -0,0 +1,745 @@
+    pub fn blocking_listen(&mut self, buffer: &mut [u8]) -> Result<Command, Error> {
+        self.setup_listen(buffer, false)?;
+        let status = self.blocking_listen_wait()?;
+        if status == Status::Write {
+            self.setup_listen_end(false)?;
+            let command = self.blocking_listen_wait_end(status)?;
+            return Ok(command);
+        }
+        Ok(Command::Read)
+    }
+
+    /// Write to an I2C master.
+    /// Returns the number of bytes written.
+    /// The buffer must have a length of at most 255 bytes on the nRF52832
+    /// and at most 65535 bytes on the nRF52840.
+    pub fn blocking_write(&mut self, buffer: &[u8]) -> Result<usize, Error> {
+        self.setup_write(buffer, false)?;
+        self.blocking_wait()
+    }
+
+    /// Same as [`blocking_write`](Twis::blocking_write) but will fail instead of copying data into RAM. Consult the module level documentation to learn more.
+    pub fn blocking_write_from_ram(&mut self, buffer: &[u8]) -> Result<usize, Error> {
+        self.setup_write_from_ram(buffer, false)?;
+        self.blocking_wait()
+    }
+
+    // ===========================================
+
+    /// Listen for commands from an I2C master with timeout.
+    ///
+    /// The buffer must have a length of at most 255 bytes on the nRF52832
+    /// and at most 65535 bytes on the nRF52840.
+    #[cfg(feature = "time")]
+    pub fn blocking_listen_timeout(&mut self, buffer: &mut [u8], timeout: Duration) -> Result<Command, Error> {
+        self.setup_listen(buffer, false)?;
+        let status = self.blocking_listen_wait_timeout(timeout)?;
+        if status == Status::Write {
+            self.setup_listen_end(false)?;
+            let command = self.blocking_listen_wait_end_timeout(status, timeout)?;
+            return Ok(command);
+        }
+        Ok(Command::Read)
+    }
+
+    /// Write to an I2C master with timeout.
+    /// Returns the number of bytes written.
+    /// See [`blocking_write`].
+    #[cfg(feature = "time")]
+    pub fn blocking_write_timeout(&mut self, buffer: &[u8], timeout: Duration) -> Result<usize, Error> {
+        self.setup_write(buffer, false)?;
+        self.blocking_wait_timeout(timeout)
+    }
+
+    /// Same as [`blocking_write`](Twis::blocking_write) but will fail instead of copying data into RAM. Consult the module level documentation to learn more.
+    #[cfg(feature = "time")]
+    pub fn blocking_write_from_ram_timeout(&mut self, buffer: &[u8], timeout: Duration) -> Result<usize, Error> {
+        self.setup_write_from_ram(buffer, false)?;
+        self.blocking_wait_timeout(timeout)
+    }
+
+    // ===========================================
+
+    /// Listen asynchronously for commands from an I2C master.
+    ///
+    /// The buffer must have a length of at most 255 bytes on the nRF52832
+    /// and at most 65535 bytes on the nRF52840.
+    pub async fn listen(&mut self, buffer: &mut [u8]) -> Result<Command, Error> {
```

Comment:
Huh, this API is very *interesting*!

- Docs could be epanded a bit. It took me a bit to figure out what "listen" means, and the fact that `buffer` is provided just in case the master does a write, but might go unused.
- It's confusing that `Command` variants are named from the master's PoV (write meaning a master->slave transfer) and the `Twis::write` method is named from the slave's PoV (write meaning a slave->master transfer).
- It could return the address (or the address index) inside Command, or maybe document "if you want to know which one it was, call `.address_match()`.

Reply:
The alternative would be to name everything from the slave's PoV but I can see that being confusing too, ugh. 

"Master wants to *read* from us so we should *write* data to it". vs "Master wants us to *write* to it so we do so"

Reply:
Hmm, to me a Command is the same regardless of PoV, since it's the "state of the bus"... But yes it might be confusing that the rx can also "write", maybe this `write` fn should be renamed to something else, like `respond`?
Is there also a better name for `listen` that would make more sense to you?

Reply:
...but if other people find reversing the Command would be more intuitive, i could totally live with that :)

Reply:
> maybe this write fn should be renamed to something else, like respond or provide?

Oooh I like this. `respond_to_read` maybe, so it's more explicit?

> Is there also a better name for listen that would make more sense to you? wait?

Noo, `listen` is great!, my confusion was due to no docs.